util/genrest/resttools/error_response.go

```diff
@@ -62,18 +76,76 @@ func GRPCToHTTP(c codes.Code) int {
 	return httpStatus
 }
 
-// ErrorResponse is a helper that formats the given response information,
-// including the HTTP Status code, a message, and any error detail types, into
-// a googleAPIError and writes the response as JSON.
-func ErrorResponse(w http.ResponseWriter, status int, message string, details ...interface{}) {
-	googleAPIError := &googleAPIError{
-		Error: &googleapi.Error{
-			Code:    status,
+// HTTPToGRPC maps the given HTTP status code to the canonical gRPC
+// Code as defined by the internal source go/http-canonical-mapping.
+func HTTPToGRPC(httpStatus int) codes.Code {
+	var gRPCCode codes.Code
+	switch {
+	case httpStatus >= 200 && httpStatus < 300:
+		gRPCCode = codes.OK
+	case httpStatus >= 300 && httpStatus < 400:
+		gRPCCode = codes.Unknown
+	case httpStatus >= 400 && httpStatus < 500:
+		gRPCCode = codes.FailedPrecondition
+	case httpStatus >= 500 && httpStatus < 600:
+		gRPCCode = codes.Internal
+	default:
+		gRPCCode = codes.Unknown
+	}
+
+	if codeOverride, ok := httpToGRPC[httpStatus]; ok {
+		gRPCCode = codeOverride
+	}
+	return gRPCCode
+}
+
+// Constants to make it obvious when we're not supplying either a gRPC or HTTP status code to
+// ErrorResponse(). The code we're not supplying will be obtained from the one we do supply.
+const (
+	NoCodeGRPC codes.Code = 9999
+	NoCodeHTTP int        = -1
+)
+
+// ErrorResponse is a helper that formats the given response information, including the HTTP or gRPC
+// status code, a message, and any error detail types, into a RestError proto message and writes the
+// response as JSON.
+func ErrorResponse(w http.ResponseWriter, httpResponseCode int, grpcStatus codes.Code, message string, details ...interface{}) {
```

Comment:
```suggestion
func ErrorResponse(w http.ResponseWriter, httpResponseCode int, grpcStatus codes.Code, message string, details ...proto.Message{}) {
```

Should we save ourselves the type cast later on while packing the `Any` and just make this a `proto.Message`?

Reply:
That's what I tried initially but it doesn't work. Quoting from our AI:  "While a single `*errdetails.ErrorInfo` is a `proto.Message`, a slice of them (`[]*errdetails.ErrorInfo`) is not a ` []proto.Message`. You must manually create the interface slice."

IOW, while a concrete message satisfies the `proto.Message` interface, a slice of concrete messages does not match a slice of that interface. (Though I'm not sure why `[]interface{}` worked before, given this. Because it's empty?)

Ref: https://go.dev/wiki/InterfaceSlice

Reply:
Also relevant: `status.Status.Details()` returns `[]any`

https://pkg.go.dev/google.golang.org/grpc@v1.71.1/internal/status#Status.Details

Reply:
Ah ok yeah that makes sense! Thanks for the in depth research there 😁